### PR TITLE
MPI: computation/communication overlap

### PR DESCRIPTION
**Pull request by FabioLuporini**
Not ready for review. 

I'm opening it only to trigger `pipelines`. 

The huge diff is mainly due to  #762

### CODE REVIEW

File: devito/dle/blocking_utils.py [hunk 1]
```diff
@@ -38,7 +39,7 @@ def fold_blockable_tree(node, exclude_innermost=False):
             if any(not is_foldable(j) for j in pairwise_folds):
                 continue
             # Maybe heuristically exclude innermost Iteration
-            if exclude_innermost is True:
+            if blockinner is False:
```
Comment (georgebisbas) [inline devito/dle/blocking_utils.py]: Does that describe a block that is inner to s-th else or does it describe the inner of a block? Or s-th else?

Comment (georgebisbas) [reply devito/dle/blocking_utils.py]: Ok, it is for loop-blocking, refers to verb. It is fine.

File: devito/cgen_utils.py [hunk 2]
```diff
@@ -158,8 +158,8 @@ def _print_ListInitializer(self, expr):
     def _print_IntDiv(self, expr):
         return expr.__str__()
 
-    def _print_Byref(self, expr):
-        return "&%s" % expr.name
+    _print_Byref = _print_IntDiv
```
Comment (mloubout) [inline devito/cgen_utils.py]: really? Looks odd to me

File: devito/cgen_utils.py [hunk 3]
```diff
@@ -158,8 +158,8 @@ def _print_ListInitializer(self, expr):
...
+    _print_IndexedPointer = _print_IntDiv
```
Comment (mloubout) [inline devito/cgen_utils.py]: same

File: devito/compiler.py [hunk 4]
```diff
@@ -379,7 +378,10 @@ def jit_compile(soname, code, compiler):
             with open(src_file, 'r') as f:
                 code = f.read()
             # Bypass the devito JIT cache
-            cache_dir = mkdtemp()
+            # Note: can't simply use Python's `mkdtemp()` as, with MPI, different
```
Comment (mloubout) [inline devito/compiler.py]: But isn't it what you want. I would assume that in some case (like in the could) each mpi rank won't have access to a common tmp dir but each to their own and will each need to compile it.

File: devito/mpi/halo_scheme.py [hunk 5]
```diff
@@ -105,10 +106,52 @@ def fmapper(self):
...
+        is (2, 2), while ``v``'s is (4, 4). The situation is depicted below.
+
+        .. code-block:: python
+
+              xx**----------------**xx     u
+            xxxx****------------****xxxx   v
+
+        Where 'x' represents a HALO point, '*' a OWNED point, and '-' a CORE point.
+        Together, '*' and '-' constitute the DOMAIN, which is guaranteed to have
+        identical size across all Functions.
```
Comment (mloubout) [inline devito/mpi/halo_scheme.py]: Is that true? I would think this is only true if all functions are not subsampled with same Dimensions.

File: devito/tools/data_structures.py [hunk 6]
```diff
@@ -188,3 +188,119 @@ def relations(self):
...
         raise NotImplementedError
+
+
+class DAG(object):
+    """
+    A trivial implementation of a directed acyclic graph (DAG).
+
+    Originally extracted from:
+
+        https://github.com/thieman/py-dag/
```
Comment (mloubout) [inline devito/tools/data_structures.py]: Is this tested?

File: devito/cgen_utils.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/cgen_utils.py]: yes because ``__str__()`` is exactly what we get to generate

File: devito/cgen_utils.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/cgen_utils.py]: same

File: devito/compiler.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/compiler.py]: unfortunately not. Think about what's going on here. The user has _manually_ modified some C file, but of course he's done that only on a single rank. All MPI processes should eventually pick up that specific file. So this actually requires some further coordination in the case of multi-node execution. Single node MPI should be fine. 

Also, with MPI, the executable ran by the MPI processes must be *exactly the same*.

File: devito/tools/data_structures.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/tools/data_structures.py]: well it's tested [here](https://github.com/thieman/py-dag/blob/master/tests/init_tests.py)

Our implementation is basically a subset of the linked one so...

File: devito/mpi/halo_scheme.py (hunk 5 above)
Comment (FabioLuporini) [reply devito/mpi/halo_scheme.py]: in fact, it's even more subtle than that, as even w/o subsampling you might have one MPI rank in each Dimension having a slightly smaller CORE region due to the number of processes not dividing evenly the grid size

But what happened here is that the docstring is misleading. What's important is that, for each FUnction, the number of 'x' and '*' is the **same on all MPI ranks**, the '-' don't matter. And this is really true. IOW, the halo/owned regions have same size across different ranks.

I'll rephrase this docstring. Thanks for pointing it out

Comment (FabioLuporini) [reply devito/mpi/halo_scheme.py]: I fixed the docstring

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**